[feat] Emitter Pools

scohen opened the pull request:
During testing at Whatnot, we found that when emitting around 50k metrics per second, the current pool implementation caused the calling processes to stall while waiting for `:gen_udp` to write data. This stalling caused significant degradation to our application, and eventually caused an outage.

To address this, we came up with the concept of using a pool of emitters. An emitter is a process that contains a socket (either UDP or domain), receives metrics and writes them to a socket. This architecture frees up the calling processes so they can write a metric and not wait on a result. Emitters also allow us to combine metrics so they fill up a UDP packet. The end result, according to our testing, moved us from 50k to 1MM metrics per second.

scohen committed:
[feat] Emitter Pools

During testing at Whatnot, we found that when emitting around 50k
metrics per second, the current pool implementation caused the calling
processes to stall while waiting for `:gen_udp` to write data. This
stalling caused significant degradation to our application, and
eventually caused an outage.

To address this, we came up with the concept of using a pool of
emitters. An emitter is a process that contains a socket (either UDP
or domain), receives metrics and writes them to a socket. This
architecture frees up the calling processes so they can write a metric
and not wait on a result. Emitters also allow us to combine metrics so
they fill up a UDP packet. The end result, according to our testing,
moved us from 50k to 780k metrics per second.

scohen commented:
I think this PR is significant enough to bump the minor version (or make this the 1.0 release). It should be backwards compatible with the old implementation, and even with the default configuration, should perform better and be safer than the pooled implementation.

scohen committed:
ignore econnrefused and eagain errors

scohen committed:
Add emit percentage metric

scohen committed:
Add emit percentage metric

Adds the emit percentage as a metric. Also added tests

scohen committed:
Change emit percent as an integer

scohen committed:
experiment: convert to buffers

scohen committed:
experiment: handle lists of buffers

scohen commented:
@arkgil PTAL!

arkgil reviewed (commented):
Thanks @scohen! I know it took a while 😅

scohen commented on lib/telemetry_metrics_statsd.ex:
```diff
@@ -284,10 +284,10 @@ defmodule TelemetryMetricsStatsd do
   exceed the Maximum Transmission Unit, or MTU, of the link, so that no data is lost on the way.
   By default the reporter will break up the datagrams at 512 bytes, but this is configurable via
-  the `:mtu` option.
+  the `:mtu` option. Properly setting the MTU will have a drastic impact on performance.
```
It's really dependent on their network configuration. You need to look at the network's MTU and account for the UDP packet overhead.

scohen replied:
BTW, the default of 512 is pretty darn small. Ethernet has a default MTU of 1500, should we increase the default to 1472?

arkgil replied:
yeah,we can bump it and cut a major. why 1472 though?

scohen replied:
1500 is the max packet size for ethernet, but IP and UDP data cut into that a bit: 
IP address headers are 20 bytes, and UDP headers are 8 

https://stackoverflow.com/questions/14993000/the-most-reliable-and-efficient-udp-packet-size

scohen commented on test/telemetry_metrics_statsd/emitter/udp_test.exs:
```diff
@@ -0,0 +1,464 @@
+      elapsed = written_at - post_emit
+      assert elapsed in 100..105
+    end
+  end
```
Can you elucidate which tests you'd like to see changed? The tests in this block test error conditions and care a lot about when things were sent. Setting up a socket and reading from it will not help me figure out when things were sent, only when they were received. 

The resolution case is also interesting, since  it tests an initial failure, then an eventual success. How should that be tested?

Similarly, the buffering tests will be fairly difficult and fiddly to test using sockets.

Can you show me a couple tests that would benefit from the approach you suggested?

arkgil replied:
I added comments on tests which I think could use without mocking. Let me know what you think.

scohen replied:
I added clarifying comments on two tests and converted the rest.

scohen commented on lib/telemetry_metrics_statsd/options.ex:
```diff
@@ -65,6 +62,39 @@ defmodule TelemetryMetricsStatsd.Options do
+        "The maximum amount of time, in milliseconds that a metric can sit in an emitter's buffer before being" <>
+          "written to the socket. Only used by the UDP emitter."
+    ],
+    emitters: [
```
I changed this to maintain performance for those people who blindly upgrade to the newest version.

 Emitters is set to a safe default, and function very differently than the previous pool. Having too many emitters hurts performance. The consideration is that if you have a large socket pool, and move to emitter pools, you will see better performance with one emitter than you saw previously. However, if we keep the old parameter, you will see a reduction of performance --especially if you had a large pool.

This is one of the reasons I suggested bumping the major version of the library when this change is landed.

arkgil replied:
Let's stick to `pool_size` and cut a major version.

arkgil replied:
I see your point regarding the upgrades, but `emitters` is a very unusual name for a pool size. Perhaps `sockets` would be a better fit?

scohen replied:
Option changed to `sockets`

scohen commented on lib/telemetry_metrics_statsd/options.ex:
```diff
@@ -65,6 +62,39 @@ defmodule TelemetryMetricsStatsd.Options do
+          "messages emitted is reduced by 50%. The percentage of messages emitted goes up by 1% if a probe message sits in" <>
+          "the queue less than the `max_queue_dwell_time`."
+    ],
+    dwell_time_check_interval: [
```
It's dependent on how often you send metrics to your backend. If you send metrics every three seconds, you can tolerate a larger dwell time, since your metrics won't get stale. Similarly, if you send every half-second, you can't tolerate a second of dwell time.

scohen commented on mix.exs:
```diff
@@ -31,7 +31,7 @@ defmodule TelemetryMetricsStatsd.MixProject do
     [
       docs: :docs,
-      dialyzer: :test,
+      dialyzer: :dev,
```
When I ran dialyzer, I received the following error:
```
Unknown error occurred: %Protocol.UndefinedError{protocol: String.Chars, value: {15, 24}, description: ""}


Legacy warning:
test/support/helpers.ex:15:24: Unknown function 'Elixir.ExUnit.Callbacks':start_supervised/1
________________________________________________________________________________
done (warnings were emitted)
Halting VM with exit status 2
```
Updating dialyzer gives me
```
test/support/helpers.ex:15:24:unknown_function
Function ExUnit.Callbacks.start_supervised/1 does not exist.
________________________________________________________________________________
done (warnings were emitted)
Halting VM with exit status 2
```
It didn't seem intentional to me to include testing code in dialyzer. Is that something you want to do?

scohen commented on mix.exs:
```diff
@@ -41,11 +41,12 @@ defmodule TelemetryMetricsStatsd.MixProject do
       {:liveness, "~> 1.0", only: :test},
-      {:mock, "~> 0.3.5", only: :test}
+      {:mock, "~> 0.3.5", only: :test},
+      {:patch, "> 0.0.0", only: :test}
```
Actually, i don't think we're using mock anywhere any more.

scohen commented on test/telemetry_metrics_statsd/emitter/congestion_test.exs:
```diff
@@ -0,0 +1,246 @@
+                 initial_emit_percentage,
+                 max_dwell_time,
+                 observed_dwell_time
+               ) > initial_emit_percentage
```
That seems like it's testing a hidden implementation detail. I'll do it, but the idea was that the percentage increases when the dwell time is below the threshold, not that it increases by a specific amount.

scohen commented on lib/telemetry_metrics_statsd/emitter/udp.ex:
```diff
@@ -0,0 +1,356 @@
+    buffer(old_buffer, data: [data, "\n" | metric_data], count: count + 1, size: total_size)
+  end
+
+  defp flush_remaining_or_incoming([], metric_data) do
```
I think you're right.

scohen commented on lib/telemetry_metrics_statsd/emitter/udp.ex:
```diff
@@ -0,0 +1,356 @@
+        :ok
+
+      {:error, :econnrefused} ->
+        # TODO: Carrying over socket lib behavior, this seems like something we'd like to know about.
```
I was referring to this: 
https://github.com/beam-telemetry/telemetry_metrics_statsd/pull/103/files#diff-a161763a2293fd06323a94c82ce22637b5f201d82e9ca4bddcefb1290a8d7477R60

arkgil replied:
@scohen would you mind removing the comment and the one below? they don't add much without the context of the original socket PR

scohen commented on lib/telemetry_metrics_statsd/emitter/domain.ex:
```diff
@@ -0,0 +1,141 @@
+  def init(%Options{} = options) do
+    # This GenServer basically shuffles messages from the calling process into the receiving socket,
+    # which appears to be the exact case for off heap message queue data.
+    Process.flag(:message_queue_data, :off_heap)
```
I read in the efficiency guide (or erlang in anger, can't remember which) that this helps with genservers that act as proxies between two sinks. Since that's what this module does, I thought it could be helpful. I did run tests and found a noticeable performance improvement when this was enabled.

> However, we still need to copy the message onto the process heap to process it and send it to the socket,

I don't know. Unlike the UDP module, there's not much processing here. A message could be flagged as "off heap" and if it's not modified, it doesn't necessarily need to be copied into this process.

scohen commented on lib/telemetry_metrics_statsd/emitter/congestion.ex:
```diff
@@ -0,0 +1,106 @@
+          min(orig_emit_percentage + @emit_percentage_increment, 1.0)
+
+        true ->
+          orig_emit_percentage
```
ah yes, this cond changed significantly, i'll change to an if.

scohen committed:
Addressed PR suggestions

scohen commented on lib/telemetry_metrics_statsd/emitter/udp.ex:
```diff
@@ -0,0 +1,356 @@
+  def handle_call({:emit, data}, _from, %__MODULE__{} = state) do
+    new_state =
+      case add_to_buffer(state, data) do
+        {:flush, buffers, new_buffer} when is_list(buffers) ->
```
the name `buffers` is giving a clue what the type that's returned is. It's a list of `buffer` records that are all ready to be flushed.

scohen commented on test/support/helpers.ex:
```diff
@@ -1,6 +1,27 @@
 
+  alias TelemetryMetricsStatsd.Options
+
+  def new_emitter(emitter_module, defaults, overrides) do
```
this was used in two different tests. It was a little convenient to have this more generic function do this for us rather than doing it twice in each test. I can change if you insist.

arkgil replied:
yeah, I'd change it

whatyouhide commented:
@arkgil @scohen is this something we're still pursuing?

arkgil commented:
It is, we’re successfully running this in production for more than a year now at 5x the scale Steve originally built it for. We’ll get this shipped soon.

On Wed, Mar 11, 2026, at 15:58, Andrea Leopardi wrote:
> *whatyouhide* left a comment (beam-telemetry/telemetry_metrics_statsd#105) <https://github.com/beam-telemetry/telemetry_metrics_statsd/pull/105#issuecomment-4039819028>
> @arkgil <https://github.com/arkgil> @scohen <https://github.com/scohen> is this something we're still pursuing?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/beam-telemetry/telemetry_metrics_statsd/pull/105#issuecomment-4039819028>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACYSVEJ22UMV3TT5MCO2QPT4QF5HVAVCNFSM6AAAAACWOMXON6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHM2DAMZZHAYTSMBSHA>.
> You are receiving this because you were mentioned.Message ID: ***@***.***>
>

arkgil commented on test/telemetry_metrics_statsd/emitter/udp_test.exs:
```diff
@@ -0,0 +1,464 @@
+  end
+
+  describe "start_link/1" do
+    test "opens an inet socket by default" do
```
```suggestion
    test "opens an ipv4 socket by default" do
```

arkgil commented on test/telemetry_metrics_statsd/emitter/udp_test.exs:
```diff
@@ -0,0 +1,464 @@
+
+      {:ok, _} = new_emitter()
+
+      assert_called :socket.open(:inet, :dgram, :udp)
+      assert_called :socket.connect(_, %{port: _, addr: {127, 0, 0, 1}, family: :inet})
```
let's open a real socket and listen on it emit a metric instead of mocking, this test is not really testing anything right now

scohen replied:
I don't think that's possible, UDP sockets don't receive any information when a connection happens.

arkgil commented on test/telemetry_metrics_statsd/emitter/udp_test.exs:
```diff
@@ -0,0 +1,464 @@
+
+      {:ok, _} = new_emitter(host: "::1", inet_address_family: :inet6)
+
+      assert_called :socket.open(:inet6, :dgram, :udp)
```
let's open a real socket and listen on it emit a metric instead of mocking, this test is not really testing anything right now

scohen replied:
Same comment as for the ipv4 test.

arkgil commented on test/telemetry_metrics_statsd/emitter/udp_test.exs:
```diff
@@ -0,0 +1,464 @@
+
+      emit(emitter, @metric)
+
+      assert_called :socket.send(_, [@metric])
```
we could test this without mocks by listetning on the receiving socket

arkgil commented on test/telemetry_metrics_statsd/emitter/udp_test.exs:
```diff
@@ -0,0 +1,464 @@
+
+      emit(emitter, @metric)
+
+      refute_any_call :socket.send()
```
we could test this without mocks by listetning on the receiving socket

arkgil commented on test/telemetry_metrics_statsd/emitter/udp_test.exs:
```diff
@@ -0,0 +1,464 @@
+      {:ok, emitter} = new_emitter(mtu: byte_size(@metric) - 1)
+      emit(emitter, @metric)
+
+      assert_called :socket.send(_, [@metric])
```
we could test this without mocks by listetning on the receiving socket

arkgil commented on test/telemetry_metrics_statsd/emitter/udp_test.exs:
```diff
@@ -0,0 +1,464 @@
+
+      emit(emitter, over_mtu_metric)
+
+      assert_called :socket.send(_, [@metric])
```
we could test this without mocks by listetning on the receiving socket

arkgil commented on test/telemetry_metrics_statsd/emitter/udp_test.exs:
```diff
@@ -0,0 +1,464 @@
+
+      refute_any_call :socket.send()
+
+      eventually(fn -> assert_called(:socket.send(_, [@metric])) end)
```
we could test this without mocks by listening on the receiving socket

arkgil commented on test/telemetry_metrics_statsd/emitter/udp_test.exs:
```diff
@@ -0,0 +1,464 @@
+        :timer.tc(fn ->
+          refute_any_call :socket.send()
+
+          rapid_eventually(fn -> assert_called :socket.send(_, [@metric]) end)
```
we could test this without mocks by listening on the receiving socket

arkgil commented on test/telemetry_metrics_statsd/emitter/udp_test.exs:
```diff
@@ -0,0 +1,464 @@
+    end
+
+    test "crashes on startup when it fails" do
+      patch(:inet, :gethostbyname, {:error, :nxdomain})
```
we could use an invalid domain name instead of mocking instead

arkgil commented on test/telemetry_metrics_statsd/emitter/udp_test.exs:
```diff
@@ -0,0 +1,464 @@
+      emit(emitter, @metric)
+
+      eventually(fn -> assert_called(Emitter.UDP.handle_info(:resolve_host, _)) end)
+      eventually(fn -> assert_called :socket.send(_, [@metric]) end)
```
we could receive on the socket instead of mocking here

arkgil commented on test/telemetry_metrics_statsd/emitter/udp_test.exs:
```diff
@@ -0,0 +1,464 @@
+      eventually(fn -> assert_called :socket.send(_, [@metric]) end)
+
+      # we should only connect once at the beginning if resolution failed.
+      assert_called :socket.connect(:socket, %{port: _, addr: {127, 0, 0, 1}, family: :inet}), 1
```
does it matter that we assert on this? after all we only care about fault tolerance here

scohen replied:
I think this is ensuring we don't repeatedly connect, IIRC, it was checking the original code's behavior.

scohen committed:
Address PR feedback

scohen committed:
Rename `emitters` option to `sockets`

scohen committed:
remove defaults and options from emitter helper

scohen commented:
> 5x the scale Steve originally built it for

To be fair, I tested it at a million metrics per second, we're not there yet ;)

arkgil commented on lib/telemetry_metrics_statsd/emitter/congestion.ex:
```diff
@@ -0,0 +1,100 @@
+defmodule TelemetryMetricsStatsd.Emitter.Congestion do
+  @moduledoc """
```
@scohen nit: we should probably `@moduledoc false` this one as we're not exposing this module to the users.

arkgil reviewed (approved):
lfg

scohen committed:
address PR feedback

scohen committed:
Return a singly wrapped error message on startup failure

scohen committed:
Fixing all versions

Normalize errors during initialization emitters via our helpers; also
add an option for starting a non-linked emitter process, which
prevents a crash on startup from taking down the test.

scohen commented:
@arkgil all green!

arkgil commented:
🚀
